docs/CodingGuidelines.md

```diff
@@ -0,0 +1,174 @@
+
+1. Ensure that all documentation is correctly referenced in the table of contents, which can be found at https://github.com/awslabs/iot-app-kit/blob/main/README.md.
+
+   In-depth guidelines on documentation: https://alpha-docs-aws.amazon.com/awsstyleguide/latest/styleguide/awsstyleguide.pdf#aws-implementation-guides. This can be a useful reference but you are not expected to read through and be 100% adherent to these guidelines.
+1. Components must have high quality testing, including integration tests which utilize the component in the way a user does.
+
+1. Tests should be written to maximize the chance that the test would actually catch a real issue, i.e. the likelihood of preventing a regression. This is the primary value of a test.  Useful talk on testing: https://www.youtube.com/watch?v=Fha2bVoC8SE3.
+
+2. Tests should also be written to minimize the pain they inflict on the development cycle. Every test represents some amount of burden, through the necessity to run it, maintain it, and alter the test as the codebase morphs, however some tests inflict more pain than others.  
+   *EXAMPLES OF PAIN*:
+  - tests failing when nothing broke
+  - flaky tests
+  - snapshot tests that vary on machine
+  - tests that require refactoring when irrelevant portions of the codebase are altered
+
+1. Tests should be written to not fail when things are not broken (this helps reduce inflicted pain). This can be achieved through techniques such as not over asserting.  
+   *GOOD ASSERTION EXAMPLE*:  
+   expect(result).toEqual(expect.objectContaining({ error: null })
+
+   *BAD ASSERTION EXAMPLE*:  
+   every test asserting on every property over and over whether or not they are relevant to the purpose of the test
+
+1. Tests names should be descriptive and from a users perspective:  
+   *GOOD NAME EXAMPLE*: “it creates new user when new user form is submitted”  
+   *BAD NAME EXAMPLE*: “button dispatches onClick event”
+
+1. Tests should, at minimum, cover all core feature use-cases and identifiable edge-cases or boundary conditions.
+
+1. Code must be written in a consistent style, utilizing the shared eslint and stylelint configurations https://github.com/awslabs/iot-app-kit/blob/main/.eslintrc.js https://github.com/awslabs/iot-app-kit/blob/main/.stylelintrc
+1. Pull requests to feature branches must contain only a single, well formatted commit that adheres to https://www.conventionalcommits.org/en/v1.0.0/
+1. Pull requests to main must have correct change logs which properly document migrations, feature changes, and most importantly breaking change.
+1. Pull requests to main must either be a single commit, or be a series of commits with prior code review plus a single commit for any necessary changes to rebase on top of main
```

Comment:
would it be possible to put an example of what this would look like?

Reply:
in the next iteration i will add a lot more detail on the git management portion and go into more detail on how to manage the commit history